### PR TITLE
queue: inline queue

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
queue: inline queue
## Testing
We can use qemu for testing.
compiling
make distclean -j20; ./tools/configure.sh -l qemu-armv8a:nsh_smp ;make -j20 running
qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic -machine virt,virtualization=on,gic-version=3 -net none -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -kernel ./nuttx

### CODE REVIEW

**Comment by anchao** [comment]
I think we should keep balance with small memory footprint devices, FYI:
https://github.com/apache/nuttx/pull/6766

**Comment by anchao** [comment]
How about implement a similar function like `list_gen`? we could use similar methods elegantly in c files or header files
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/list_gen.h
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/slist.h

**Comment by xiaoxiang781216** [comment]
> How about implement a similar function like `list_gen`? we could use similar methods elegantly in c files or header files https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/list_gen.h https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/slist.h

but, zephyr approach is similar as https://github.com/apache/nuttx/blob/master/include/nuttx/list.h. Why not continue use list.h?

**Comment by anchao** [comment]
How about using macros to isolate different implementations?
list.h:
```
#ifdef INLINE_QUEUE
#define STATIC_INLINE static line
#else
#define STATIC_INLINE
#endif

STATIC_INLINE void add_queue(void);
#ifdef INLINE_QUEUE
STATIC_INLINE void add_queue(void)
{
 #implement
}
#endif
```

list.c:
```
#undef INLINE_QUEUE
#include <list.h>
```

Make.defs
```
ifneq ($(INLINE_QUEUE),y)
CSRCS += list.c
endif
```

**Comment by pkarashchenko** [comment]
Do we have any details on this change? Like performance vs size measurements. What is the value that we are trying to achieve with this change?

**Comment by anchao** [comment]
> How about using macros to isolate different implementations? list.h:
> 
> ```
> #ifdef INLINE_QUEUE
> #define STATIC_INLINE static line
> #else
> #define STATIC_INLINE
> #endif
> 
> STATIC_INLINE void add_queue(void);
> #ifdef INLINE_QUEUE
> STATIC_INLINE void add_queue(void)
> {
>  #implement
> }
> #endif
> ```
> 
> list.c:
> 
> ```
> #undef INLINE_QUEUE
> #include <list.h>
> ```
> 
> Make.defs
> 
> ```
> ifneq ($(INLINE_QUEUE),y)
> CSRCS += list.c
> endif
> ```

@hujun260 How about this approach? You could still merge all list implementations into the header file

**Comment by hujun260** [comment]
> > How about using macros to isolate different implementations? list.h:
> > ```
> > #ifdef INLINE_QUEUE
> > #define STATIC_INLINE static line
> > #else
> > #define STATIC_INLINE
> > #endif
> > 
> > STATIC_INLINE void add_queue(void);
> > #ifdef INLINE_QUEUE
> > STATIC_INLINE void add_queue(void)
> > {
> >  #implement
> > }
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > list.c:
> > ```
> > #undef INLINE_QUEUE
> > #include <list.h>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Make.defs
> > ```
> > ifneq ($(INLINE_QUEUE),y)
> > CSRCS += list.c
> > endif
> > ```
> 
> @hujun260 How about this approach? You could still merge all list implementations into the header file

If we use this approach, we would need to write the same implementation in both the header file and the source file. Subsequent modifications could lead to omissions.

**Comment by anchao** [comment]
> 
> If we use this approach, we would need to write the same implementation in both the header file and the source file. Subsequent modifications could lead to omissions.

You only need to keep one implementation in the header file, and the C source file needs to include the header file without implementing the function

File: include/nuttx/queue.h [hunk 1]
```diff
@@ -31,6 +31,14 @@
  * Pre-processor Definitions
  ****************************************************************************/
 
+#ifdef CONFIG_LIBC_INLINE_QUEUE
+#  ifndef STATIC_INLINE
+#define STATIC_INLINE static inline_function
```
Comment (anchao) [inline include/nuttx/queue.h:36]:
```suggestion
#    define STATIC_INLINE static inline_function
```

File: libs/libc/queue/CMakeLists.txt [hunk 2]
```diff
@@ -18,15 +18,6 @@
...
-          sq_remlast.c
-          sq_remfirst.c
-          sq_remafter.c
-          sq_count.c
-          dq_addafter.c
-          dq_remlast.c
-          dq_remfirst.c
-          dq_remafter.c
-          dq_count.c)
+if(NOT CONFIG_LIBC_INLINE_QUEUE)
```
Comment (anchao) [inline libs/libc/queue/CMakeLists.txt:21]: remove the check

File: libs/libc/queue/Make.defs [hunk 3]
```diff
@@ -18,12 +18,15 @@
 #
 ############################################################################
 
+ifneq ($(CONFIG_LIBC_INLINE_QUEUE),y)
```
Comment (anchao) [inline libs/libc/queue/Make.defs:21]: ditto

File: libs/libc/queue/queue.c [hunk 4]
```diff
@@ -19,36 +19,18 @@
  ****************************************************************************/
 
 /****************************************************************************
- * Included Files
+ * Pre-processor Definitions
```
Comment (anchao) [inline libs/libc/queue/queue.c:22]:
```
/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/config.h>

/****************************************************************************
 * Pre-processor Definitions
 ****************************************************************************/

#ifndef CONFIG_LIBC_INLINE_QUEUE

#  define CONFIG_LIBC_INLINE_QUEUE
#  define STATIC_INLINE

/****************************************************************************
 * Included Files
 ****************************************************************************/

#include <nuttx/queue.h>

/****************************************************************************
 * Public Functions
 ****************************************************************************/

#endif

```

File: libs/libc/queue/Kconfig [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+config LIBC_INLINE_QUEUE
+	bool "Inline queue list to improve the performance"
+	default n
```
Comment (xiaoxiang781216) [inline libs/libc/queue/Kconfig:8]: !DEFAULT_SMALL

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**